force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,9 +181,33 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
+        if(accounts.isEmpty()){
+            Account newAcc = new Account(Name = accountName);
+            Database.insert(newAcc);
+            accounts.add(newAcc);
+        } 
+        List<Opportunity> opportunitiesThatExist = [SELECT Name, Id FROM Opportunity WHERE Name IN :oppNames AND AccountId = :accounts[0].Id];
```

Comment:
By filtering by AccountId here you are excluding opportunities that already exist, but are missing the related Account